metabrainz/musicbrainz-server
# MBS-4685 / MBS-13084: Allow modifying and removing edit notes

### Implement MBS-4685 / MBS-13084

# Problem
There is no way for users to delete or fix their edit notes. While obviously this is not desirable all the time, it would be great to have a way for editors to delete or change an edit note if they, say, immediately notice they pasted the wrong thing (even more so if it's for example private data).

There's also no way for *admins* to delete or amend edit notes, which means notes with spam such as [this one](https://musicbrainz.org/edit/96631022#note-96631022-1) cannot be acted on.

# Solution
This adds a new table `edit_note_change` to track changes to edit notes.

Admins can edit or delete an edit note at any time of their choosing. For anyone else, the following conditions must apply:
1) the edit note is by the user (obviously)
2) the edit note hasn't already been deleted
3) the edit note isn't older than 24 hours
4) the edit note hasn't gotten a reply by anyone else than the editor or ModBot
5) the user isn’t disabled to add edit notes (by admin)

If the edit note is too old or has a reply, but the user really wants to edit it (say, someone answered to say "Is that your password? Whoops!"), then we tell them to contact us so an admin can make the necessary changes if they agree with the reasoning.

![Screenshot from 2021-12-08 16-39-22](https://user-images.githubusercontent.com/1069224/145383826-973143fe-efc7-40a0-85a6-91a0a8f40043.png)

# Testing
Manually, plus wrote a few tests.

Tested manually: 
* Added notes to stuff, tried removing them with a reason, without a reason, editing with a reason, without a reason, editing and deleting them as admin.

Covered in tests:
* The expected amount of modify note controls are shown for a non-admin editor and for an admin (so, the expected amount of notes are editable).
* Edit notes cannot be modified by non-admin non-owner even by navigating to the form directly.
* Edit notes with replies cannot be modified by non-admin owner even by navigating to the form directly.
* Old edit notes cannot be modified by non-admin owner even by navigating to the form directly (limit difference is 24 hours, but tested difference is many years).
* Edit notes cannot be modified by non-admin owner when their edit note privs are off, even by navigating to the form directly.
* Edit owner can modify edit note even if themselves and ModBot have left further notes, as long as nobody else has.
* Actually modifying the edit note as the owner works, and after modification the reason and date of the modification are listed.
* Actually deleting the edit note as the owner works, and after deletion the reason of the deletion is listed.
* Admins can delete old edit notes.
* Admins can "delete" already deleted edit notes to change the reason (for example if the deleting editor left an insult there).
* Actually deleting the edit note as an admin works, and after deletion the reason of the deletion (or the fact there was no reason) is listed.
* Edit notes cannot be edited to a blank string, or a string devoid of content like "a" (as with required edit note additions).
* Already deleted edit notes cannot be deleted again nor edited by their owner.
